File: noisecluster-jvm/src/test/scala/noisecluster/test/specs/transport/udp/UnicastUdpSpec.scala
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
package noisecluster.test.specs.transport.udp

import akka.actor.ActorSystem
import noisecluster.jvm.test.utils._
import noisecluster.jvm.transport.udp.{Defaults, Source, Target}
import org.scalatest.{FlatSpec, Matchers}

import scala.concurrent.{ExecutionContext, Future}
import scala.util.Random

class UnicastUdpSpec extends FlatSpec with Matchers {
  private var testDataSent = 0L
  private var testDataReceived = 0L

  private def testDataHandler: (Array[Byte], Int) => Unit = (_: Array[Byte], length: Int) => {
    testDataReceived += length
  }

  private implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
  private implicit val loggingSystem = ActorSystem("testLoggingSystem")
  private val address = "localhost"
  private val sourcePort = 49142
  private val targetPort_01 = 49143
  private val targetPort_02 = 49144
  private val targetPort_03 = 49145

  private val source: Source = Source(Seq((address, targetPort_01), (address, targetPort_02), (address, targetPort_03)), sourcePort)
  private val target_01: Target = Target(address, targetPort_01, Defaults.BufferSize)
  private val target_02: Target = Target(address, targetPort_02, Defaults.BufferSize)
  private val target_03: Target = Target(address, targetPort_03, Defaults.BufferSize)

  private val testByteArraySize = 1000

  private var targetFuture_01 = Future {
    target_01.start(testDataHandler)
  }

  private val targetFuture_02 = Future {
    target_02.start(testDataHandler)
  }

  private val targetFuture_03 = Future {
    target_03.start(testDataHandler)
  }

  waitUntil(what = "target becomes active", waitTimeMs = 500, waitAttempts = 10) {
    target_01.isActive && target_02.isActive && target_03.isActive
  }

  "A source and a target" should "successfully exchange data" in {
    val bytes = Array.ofDim[Byte](testByteArraySize)
    Random.nextBytes(bytes)
    source.send(bytes)
    testDataSent += testByteArraySize

    waitUntil(what = "data is received by target", waitTimeMs = 500, waitAttempts = 10) {
      testDataSent * 3 == testDataReceived
    }

    testDataSent * 3 should be(testDataReceived)
    testDataReceived should be(testByteArraySize * 3)
  }

  "A target" should "successfully stop accepting data" in {
    target_01.isActive should be(true)
    target_02.isActive should be(true)
    target_03.isActive should be(true)
    target_01.stop()

    waitUntil(what = "target becomes inactive", waitTimeMs = 500, waitAttempts = 10) {
      !target_01.isActive
    }

    target_01.isActive should be(false)

    assertThrows[IllegalStateException] {
      target_01.stop()
    }
  }

  it should "successfully restart and accept data" in {
    target_01.isActive should be(false)
    targetFuture_01 = Future {
      target_01.start(testDataHandler)
    }

    waitUntil(what = "target becomes active", waitTimeMs = 500, waitAttempts = 10) {
      target_01.isActive
    }

    target_01.isActive should be(true)

    val bytes = Array.ofDim[Byte](testByteArraySize)
    Random.nextBytes(bytes)
    source.send(bytes)
    testDataSent += testByteArraySize

    waitUntil(what = "data is received by target", waitTimeMs = 500, waitAttempts = 10) {
      testDataSent * 3 == testDataReceived
    }

    testDataSent * 3 should be(testDataReceived)
    testDataReceived should be(testByteArraySize * 6)
  }

  it should "fail to close an active connection" in {
    assertThrows[IllegalStateException] {
      target_01.close()
    }
  }

  it should "successfully stop and close its connection" in {
    target_01.isActive should be(true)
    target_02.isActive should be(true)
    target_03.isActive should be(true)
    target_01.stop()
    target_02.stop()
    target_03.stop()
    target_01.close()
    target_02.close()
    target_03.close()
    target_01.isActive should be(false)
    target_02.isActive should be(false)
    target_03.isActive should be(false)
  }
}

File: noisecluster-win/noisecluster/win/test/transport/udp/UnicastUdpSpec.cs
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using log4net.Config;
using noisecluster.win.transport;
using noisecluster.win.transport.udp;
using NUnit.Framework;

namespace noisecluster.win.test.transport.udp
{
    [TestFixture]
    public class UnicastUdpSpec
    {
        private long _testDataSent;
        private long _testDataReceived;

        private readonly DataHandler _testDataHandler;

        private readonly Source _source;
        private readonly Target _target01;
        private readonly Target _target02;
        private readonly Target _target03;

        private readonly int _testByteArraySize;

        private readonly Random _rnd;
        private Task _targetTask01;
        private Task _targetTask02;
        private Task _targetTask03;

        public UnicastUdpSpec()
        {
            BasicConfigurator.Configure();
            _testDataSent = 0;
            _testDataReceived = 0;

            _testDataHandler = (data, length) => { _testDataReceived += length; };

            const string address = "127.0.0.1";
            const int sourcePort = 49142;
            const int targetPort01 = 49143;
            const int targetPort02 = 49144;
            const int targetPort03 = 49145;

            _source = new Source(
                new List<Tuple<string, int>>
                {
                    Tuple.Create(address, targetPort01),
                    Tuple.Create(address, targetPort02),
                    Tuple.Create(address, targetPort03)
                },
                sourcePort
            );

            _target01 = new Target(targetPort01);
            _target02 = new Target(targetPort02);
            _target03 = new Target(targetPort03);

            _testByteArraySize = 1000;

            _rnd = new Random();
        }

        [Test]
        public void T01_SourceAndTarget_should_ExchangeData()
        {
            _targetTask01 = new Task(() => { _target01.Start(_testDataHandler); });
            _targetTask02 = new Task(() => { _target02.Start(_testDataHandler); });
            _targetTask03 = new Task(() => { _target03.Start(_testDataHandler); });
            _targetTask01.Start();
            _targetTask02.Start();
            _targetTask03.Start();

            Utils.WaitUntil("target becomes active", 500, 10, () => _target01.IsActive());
            Utils.WaitUntil("target becomes active", 500, 10, () => _target02.IsActive());
            Utils.WaitUntil("target becomes active", 500, 10, () => _target03.IsActive());

            var bytes = new byte[_testByteArraySize];
            _rnd.NextBytes(bytes);
            _source.Send(bytes);
            _testDataSent += _testByteArraySize;

            Utils.WaitUntil("data is received by target", 500, 10, () => _testDataSent * 3 == _testDataReceived);

            Assert.IsTrue(_testDataSent * 3 == _testDataReceived);
            Assert.IsTrue(_testDataReceived == _testByteArraySize * 3);
        }

        [Test]
        public void T02_Target_should_StopAcceptingData()
        {
            Assert.IsTrue(_target01.IsActive());
            Assert.IsTrue(_target02.IsActive());
            Assert.IsTrue(_target03.IsActive());
            _target01.Stop();

            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target01.IsActive());

            Assert.IsFalse(_target01.IsActive());

            Assert.Throws<InvalidOperationException>(() => { _target01.Stop(); });
        }

        [Test]
        public void T03_Target_should_RestartAcceptingData()
        {
            Assert.IsFalse(_target01.IsActive());

            _targetTask01 = new Task(() => { _target01.Start(_testDataHandler); });
            _targetTask01.Start();

            Utils.WaitUntil("target becomes active", 500, 10, () => _target01.IsActive());

            Assert.IsTrue(_target01.IsActive());

            var bytes = new byte[_testByteArraySize];
            _rnd.NextBytes(bytes);
            _source.Send(bytes);
            _testDataSent += _testByteArraySize;

            Utils.WaitUntil("data is received by target", 500, 10, () => _testDataSent * 3 == _testDataReceived);

            Assert.IsTrue(_testDataSent * 3 == _testDataReceived);
            Assert.IsTrue(_testDataReceived == _testByteArraySize * 6);
        }

        [Test]
        public void T04_Target_should_FailToCloseActiveConnection()
        {
            Assert.Throws<InvalidOperationException>(() => { _target01.Close(); });
        }

        [Test]
        public void T05_Target_should_StopAndCloseItsConnection()
        {
            Assert.IsTrue(_target01.IsActive());
            Assert.IsTrue(_target02.IsActive());
            Assert.IsTrue(_target03.IsActive());

            _target01.Stop();
            _target02.Stop();
            _target03.Stop();

            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target01.IsActive());
            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target02.IsActive());
            Utils.WaitUntil("target becomes inactive", 500, 10, () => !_target03.IsActive());

            _target01.Close();
            _target02.Close();
            _target03.Close();

            Assert.IsFalse(_target01.IsActive());
            Assert.IsFalse(_target02.IsActive());
            Assert.IsFalse(_target03.IsActive());
        }
    }
}